### PR TITLE
Perform Search Result AJAX on scroll

### PR DESCRIPTION
**Pull request by crhallberg**
Using [Hunt](https://github.com/jeremenichelli/hunt), make AJAX calls when search results are in view.

- [x] Check Item Statuses
- [x] Check Save Statuses
- [x] Optimize checking all item statuses at once
- [x] Optimize checking all save statuses at once
- [x] Fix problem with save statuses failing to load
- [x] Fix problem with item statuses getting intermittently dropped
- [x] Consider adding this functionality for automatic OpenURL loading as well.

### CODE REVIEW

**Comment by crhallberg** [comment]
I think this is ready to go. @EreMaijala if you wanted to take a look, I would value your advice.

File: module/VuFind/src/VuFind/Controller/AjaxController.php [hunk 1]
```diff
@@ -361,6 +361,82 @@ protected function reduceServices(array $services)
...
+     * Get Item Statuses
+     *
+     * This is responsible for printing the holdings information for a
+     * collection of records in JSON format.
+     *
+     * @return \Zend\Http\Response
+     * @author Chris Delis <cedelis@uillinois.edu>
+     * @author Tuan Nguyen <tuan@yorku.ca>
+     */
+    protected function getSingleItemStatusAjax()
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Controller/AjaxController.php]: This seems like a lot of duplicated code for what should just be an edge case of the existing logic. Do we really need this? Can't we just send an array containing one ID to the existing logic?

File: module/VuFind/src/VuFind/Controller/AjaxController.php [hunk 2]
```diff
@@ -528,6 +604,46 @@ protected function getItemStatusGroup($record, $messages, $callnumberSetting)
      *
      * @return \Zend\Http\Response
      */
+    protected function getSingleSaveStatusAjax()
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Controller/AjaxController.php]: Ditto regarding redundant code.

**Review by demiankatz — COMMENTED** [review]
Sorry if I'm missing something, but I'd really like to see if we can reduce the amount of copy-and-paste coding in the controller before taking a deeper look at this. I'm hoping that we actually don't need these new AJAX methods at all.

A related thought: I still think there might be some benefit to queuing IDs instead of triggering a separate AJAX call for each ID, so that the algorithm for the observer is something like:

- Add ID to retrieval queue
- If retrieval timer does not exist, start it; if it exists, reset the clock.
- When retrieval timer runs out, run a single AJAX call to retrieve results for all queued IDs.

I know that we discussed this at least briefly, and I'm sure that performance characteristics will vary based on a number of variables... but my feeling is that for many ILSes, it's going to be less expensive to do one accumulated operation instead of many parallel ones... and in some cases, the parallel approach could actively cause problems by significantly increasing simultaneous connection counts to limited resources.

I admit that I haven't looked very closely at the Javascript yet, so maybe more is accounted for there than I'm giving credit for... but I wanted to get these issues in writing first, especially because in a way they go hand-in-hand: if we go with a queueing approach, then we have even less of a need for special case one-ID-only AJAX routines.

**Comment by EreMaijala** [comment]
We are already doing more or less what this PR does, though with jquery.inview plugin (I can't comment how it compares to Hunt), and I'm leaning on agreeing with Demian that a retrieval queue that would reduce the number of JS calls would make sense. Disregarding any performance implications related to the ILS's, the problem with fetching a single result per AJAX request is that the browser limit of simultaneous AJAX requests is quickly reached if the user scrolls or pages through the results quickly enough, and that results in the latest requests to be queued while the possible-now-again-hidden ones are retrieved. Then again, if the retrieval of a single result is slow for some reason, it doesn't block other results from being retrieved, so there are positives in both approaches.

While I haven't tested it at all, it might be useful to try a hybrid approach such as
1.) Retrieve the results that are displayed initially in a single call.
2.) Retrieve more results one by one (or in small batches?) but only after they've been visible for at least x milliseconds or scrolling has stopped.

In any case if a batch retrieval is used, it must not fail completely e.g. if one ILS cannot be contacted.

**Comment by crhallberg** [comment]
What is the state of this PR? I made the queuing changes that Ere suggested, which completes my checklist.

**Comment by demiankatz** [comment]
I think it's back in my court for review. I'll try to take a look before too long.

**Comment by demiankatz** [comment]
Is #979 going to have any impact on this? Should we get that one merged, and then merge master into this, before proceeding?

**Comment by crhallberg** [comment]
It might be best to merge this one first and then #979, but they shouldn't cause conflicting behaviour.

**Comment by demiankatz** [comment]
@crhallberg, I'm not sure what is going wrong exactly, but this is not working for me at all. If I switch to the hunt branch, none of my item statuses show up. If I switch back to master, everything works. Perhaps there is some specific combination of display settings that is problematic... but this is definitely not ready to merge yet.

**Comment by crhallberg** [comment]
Are there steps to recreate the dropped item issue?

**Comment by demiankatz** [comment]
@crhallberg, I believe that I was testing with my real VoyagerRestful.ini in place to ensure consistent results, but I don't remember it being difficult to replicate the issue. Let me know if/when you need me to take another look.

**Comment by demiankatz** [comment]
@crhallberg, just re-tested this, and I'm still seeing the same problem. Here's a simple, easy-to-reproduce test:

1.) Set your ILS driver to "Sample" -- this causes every record to have the exact same status all of the time.
2.) Perform a search on master
3.) Perform the same search on the hunt branch in a different tab
4.) Scroll down both result sets and compare side-by-side; the hunt version always seems to be missing some statuses.

**Comment by crhallberg** [comment]
This is ready for another test!

**Comment by demiankatz** [comment]
@crhallberg, this is looking pretty good to me, and all automated tests are passing. I'm going to try to put it through a couple more manual rounds, but I think it should be clear to go out in release 4.1.

@EreMaijala, any last-minute thoughts? This is now using a queue, so I think it meets the basic requirements... but another set of eyes wouldn't hurt.

**Comment by demiankatz** [comment]
@crhallberg / @EreMaijala -- I noticed that as implemented, the AJAX calls were going out in parallel, and I had some concern that this might overload some ILS drivers. I made some changes in commit https://github.com/vufind-org/vufind/pull/950/commits/bb4f2a253dbe0b6a5d7fd5b0880536a751ed715a to hopefully serialize the requests. It seems to be working fine, but I'd appreciate your comments on this approach in case I've overlooked anything important.

**Comment by demiankatz** [comment]
Depending on your reactions to my change, that also opens up the question of whether to do the same thing for save statuses....

Also, two other unrelated questions:

1.) Has anyone tested this in old versions of IE yet? It works for me in Chrome and Firefox but I haven't taken time to try MS products yet.

2.) Should we consider using a dependency manager like bower to populate our JS vendor directory in future? I'm just wondering if there's a better way to keep track of which versions of which things we're using. That's a longer-term question, of course, and doesn't really affect this PR -- it's just that adding hunt to the vendor directory made me think of it. It would be nice to have an easy way of detecting upgrades to libraries we're using.

**Comment by EreMaijala** [comment]
I haven't tested this yet, but there are still a couple of issues as far as I can see:

1. Shouldn't ajax-pending class be added already when a lookup is queued? I can't see any load indicators at all.
2. I don't think my previous comment "In any case if a batch retrieval is used, it must not fail completely e.g. if one ILS cannot be contacted." has been addressed, right? I think that it's important when doing a batch lookup to not fail the whole request if one status lookup fails.
3. It would be a good idea to do the same batching of requests for save status lookups too. 
4. While at it, why not do it also for OpenURL lookups? They currently have two modes: lookup on click or automatic lookup when results are displayed. I think it should (optionally or by default) work the same way as status lookup.

**Comment by demiankatz** [comment]
@EreMaijala, regarding your point 2, is this an existing problem in the current system (where everything is loaded simultaneously), or are you seeing it as a new problem related to the revised system here? We'll have to figure out a good way to set up a test scenario to check this out....

Regarding point 3, I think save statuses should already be getting batched here, though they're not being forced into a serial queue the same way as item statuses, because I didn't want to make that change everywhere before Chris had reviewed my changes to item statuses.

Regarding OpenURL lookups, that should be possible to set up too... though I think we should make sure we're happy with the existing logic before we add more.

File: themes/bootstrap3/js/check_save_statuses.js [hunk 3]
```diff
@@ -1,58 +1,117 @@
...
-      return null;
+var saveStatusObjs = [];
+var saveStatusEls = {};
+var saveStatusTimer = null;
+var saveStatusDelay = 200;
+var saveStatusRunning = false;
+
+function runSaveAjaxForQueue() {
+  // Only run one save status AJAX request at a time:
+  if (saveStatusRunning) {
```
Comment (demiankatz) [inline themes/bootstrap3/js/check_save_statuses.js]: I see that you've added saveStatusRunning, but it looks like you are not clearing it. I think it needs to get set to false in the .done and .fail callbacks of the AJAX routine or else only the first batch will load and then we'll go into an infinite wait loop.

**Comment by EreMaijala** [comment]
@demiankatz, Yes, sorry, I forgot the default behavior is to load everything in a single batch since we moved to a one-by-one method a good while ago. So yes, my point 2 is a problem in the existing system too.

**Comment by demiankatz** [comment]
@EreMaijala, in that case, it seems that there are two possible approaches to this:

1.) Make the AJAX handler more error-tolerant, which is a good idea but may be complex (especially if the error tolerance actually needs to be built into the MultiBackend driver -- which seems like it might be the case).

2.) Add a configuration option to allow item statuses to be configured to run on a one-by-one basis, which would replicate your existing behavior.

Any thoughts/preferences?

**Comment by EreMaijala** [comment]
I think I'd prefer option 1. MultiBackend is currently just doing a one-by-one fetch, but it should really group the records by source and then call getStatuses for all groups. I think we could define a request status code field or something in the response where MultiBackend could put any exception messages it would get from the backends. All this, I realise now, can be a completely separate task, though, so I'm ok with continuing with the current approach in this PR.

**Comment by demiankatz** [comment]
@EreMaijala, sounds good to me. Would you like to open a JIRA ticket about this task so we don't lose track of it? Let me know if I can be of any further assistance.

@crhallberg, with this complex issue off the table (or at least, off this part of the table), it seems like we might realistically be able to finish this up in time for 4.1. Note my review above, though. Maybe we can start a separate PR for the OpenURL piece after this one is done to avoid getting distracted at this stage of the game.

**Comment by EreMaijala** [comment]
https://vufind.org/jira/browse/VUFIND-1245 created. I wrote down another approach which would first try getStatuses and then resort to multiple getStatus calls if the former threw an exception.

**Comment by crhallberg** [comment]
Is there way to make such an error occur, so I can see what my options for robustness are?

**Comment by demiankatz** [comment]
@crhallberg, you could probably adjust the Demo driver so that it throws an exception in getStatuses when it encounters one particular record ID. But I think most of the adjustments are going to be on the server side rather than the client -- and that's why I think we can move forward with finishing this up, and leave the robustness issue to VUFIND-1245. Right now, I'm most concerned with fixing the save status queuing bug mentioned above (at least, I'm pretty sure it's a bug -- haven't had time to test yet).

**Comment by crhallberg** [comment]
I added openurl support for when openurls are set to autoload. Looking into failures now.

**Comment by demiankatz** [comment]
@crhallberg, my main priority is to just stabilize what we have so we can merge it on Monday. See my note above about save statuses being broken by recent changes. Please don't worry about better failure handling now -- we can deal with that later. I think this is very close to being stable, and so I just want to get it merged so it can go out in the release. We can add more stuff later. :-)

Thanks!

**Comment by demiankatz** [comment]
@crhallberg, I think this is close to being fully functional, but as @EreMaijala mentioned earlier, the Hunt version of the code seems to be suppressing the "Loading..." messages prematurely. If I use the master branch, I see "Loading..." in several places while the AJAX call processes, and then these are replaced by the final status information. In the hunt branch, I never see these. As I scroll down, statuses gradually pop into place, but there is no indication that they are coming, and it's hard to tell whether or not the loading process is finished at any given point. I saw you made some adjustments to the use of .ajax-pending, but there must be some other factor in play as well. Any ideas?

**Pull request merged**